### ISSUE TITLE
Multiline strings contains extra whitespaces

### ISSUE THREAD

**Issue by tarastriska** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#888]
### Steps to reproduce
1. Seed data with string contains end of line (\r\n)
2. Use .net core entity framework tool for migration and add new migration
3. Generate sql with _**Script-Migration -i**_
4. Generated script contains extra symbols because of tabulation added.

### The issue
Please excuse me if this is not related to Pomelo. I will open it in EF core project in that case.
The issue is with multi-line strings when I seed some data. Below is example

```
DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = 'someId') THEN

    UPDATE `SomeTable` SET `SomeColumn` = 'Some multiline
    String. That Should
    Not Contain whitepaces. But it contains because of tabulation'
    WHERE SomeCondition;
    SELECT ROW_COUNT();


    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;
```

### Further technical details

MySQL version: 5.6
Pomelo.EntityFrameworkCore.MySql version: 2.1.4
Microsoft.AspNetCore.App version: 2.1

**Comment by lauxjpn** [comment]
This seems to be an issue in EF Core upstream.

I used the following code to test this:

```c#
using System.Diagnostics;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace IssueConsoleTemplate
{
    public class Blog
    {
        public int BlogId { get; set; }
        public string Description { get; set; }
    }

    public class Issue888Context : DbContext
    {
        public DbSet<Blog> Blogs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                .UseMySql("server=127.0.0.1;port=3308;user=root;password=;database=Issue888")
                .EnableDetailedErrors();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            // Debugger.Launch();
            modelBuilder.Entity<Blog>()
                .HasData(new Blog { BlogId = 1, Description = "This text\r\nuses\r\n3 lines!" });
        }
    }

    internal class Program
    {
        private static void Main()
        {
            using var context = new Issue888Context();

            var theBlog = context.Blogs.Single();

            Debug.Assert(string.Equals(theBlog.Description, "This text\r\nuses\r\n3 lines!"));
        }
    }
}
```

I then executed the following command to generate a new migration:

```c#
dotnet ef migrations add MigrationWithMultilineSeeding --verbose
```

The following migration file was generated, which is still correct:

```c#
using Microsoft.EntityFrameworkCore.Metadata;
using Microsoft.EntityFrameworkCore.Migrations;

namespace IssueConsoleTemplate.Migrations
{
    public partial class MigrationWithMultilineSeeding : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "Blogs",
                columns: table => new
                {
                    BlogId = table.Column<int>(nullable: false)
                        .Annotation("MySql:ValueGenerationStrategy", MySqlValueGenerationStrategy.IdentityColumn),
                    Description = table.Column<string>(nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_Blogs", x => x.BlogId);
                });

            migrationBuilder.InsertData(
                table: "Blogs",
                columns: new[] { "BlogId", "Description" },
                values: new object[] { 1, @"This text
uses
3 lines!" });
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropTable(
                name: "Blogs");
        }
    }
}
```

I then executed the following command the create an idempotent script:

```
dotnet ef migrations script --idempotent
```

The following SQL script was generated, which contains the incorrect `INSERT` value due to the automatic indentation EF Core applies:

```SQL
CREATE TABLE IF NOT EXISTS `__EFMigrationsHistory` (
    `MigrationId` varchar(95) NOT NULL,
    `ProductVersion` varchar(32) NOT NULL,
    CONSTRAINT `PK___EFMigrationsHistory` PRIMARY KEY (`MigrationId`)
);


DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = '20191023132216_MigrationWithMultilineSeeding') THEN

    CREATE TABLE `Blogs` (
        `BlogId` int NOT NULL AUTO_INCREMENT,
        `Description` longtext NULL,
        CONSTRAINT `PK_Blogs` PRIMARY KEY (`BlogId`)
    );

    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;


DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = '20191023132216_MigrationWithMultilineSeeding') THEN

    INSERT INTO `Blogs` (`BlogId`, `Description`)
    VALUES (1, 'This text
    uses
    3 lines!');

    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;


DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = '20191023132216_MigrationWithMultilineSeeding') THEN

    INSERT INTO `__EFMigrationsHistory` (`MigrationId`, `ProductVersion`)
    VALUES ('20191023132216_MigrationWithMultilineSeeding', '3.0.1-dev');

    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;
```

In EF Core, the following code indents the generated statements in case an idempotent script is being generated:

https://github.com/aspnet/EntityFrameworkCore/blob/a1cb74bd60522d6ef3a31df3aab43bc2d8586601/src/EFCore.Relational/Migrations/Internal/Migrator.cs#L364-L377

```c#
if (idempotent)
{
    builder.AppendLine(_historyRepository.GetBeginIfNotExistsScript(migration.GetId()));
    using (builder.Indent())
    {
        builder.AppendLines(command.CommandText);
    }


    builder.AppendLine(_historyRepository.GetEndIfScript());
}
else
{
    builder.AppendLine(command.CommandText);
}
```

That means, that the same issue happens for all database providers, including SQL Server:

```sql
IF OBJECT_ID(N'[__EFMigrationsHistory]') IS NULL
BEGIN
    CREATE TABLE [__EFMigrationsHistory] (
        [MigrationId] nvarchar(150) NOT NULL,
        [ProductVersion] nvarchar(32) NOT NULL,
        CONSTRAINT [PK___EFMigrationsHistory] PRIMARY KEY ([MigrationId])
    );
END;

GO

CREATE TABLE [Blogs] (
    [BlogId] int NOT NULL IDENTITY,
    [Description] nvarchar(max) NULL,
    CONSTRAINT [PK_Blogs] PRIMARY KEY ([BlogId])
);

GO

IF EXISTS (SELECT * FROM [sys].[identity_columns] WHERE [name] IN (N'BlogId', N'Description') AND [object_id] = OBJECT_ID(N'[Blogs]'))
    SET IDENTITY_INSERT [Blogs] ON;
INSERT INTO [Blogs] ([BlogId], [Description])
VALUES (1, N'This text
uses
3 lines!');
IF EXISTS (SELECT * FROM [sys].[identity_columns] WHERE [name] IN (N'BlogId', N'Description') AND [object_id] = OBJECT_ID(N'[Blogs]'))
    SET IDENTITY_INSERT [Blogs] OFF;

GO

INSERT INTO [__EFMigrationsHistory] ([MigrationId], [ProductVersion])
VALUES (N'20191023140106_MigrationWithMultilineSeedingMssql', N'3.0.0');

GO


PS E:\Sources\Issue888Mssql\IssueConsoleTemplate> dotnet ef migrations script --idempotent
IF OBJECT_ID(N'[__EFMigrationsHistory]') IS NULL
BEGIN
    CREATE TABLE [__EFMigrationsHistory] (
        [MigrationId] nvarchar(150) NOT NULL,
        [ProductVersion] nvarchar(32) NOT NULL,
        CONSTRAINT [PK___EFMigrationsHistory] PRIMARY KEY ([MigrationId])
    );
END;

GO

IF NOT EXISTS(SELECT * FROM [__EFMigrationsHistory] WHERE [MigrationId] = N'20191023140106_MigrationWithMultilineSeedingMssql')
BEGIN
    CREATE TABLE [Blogs] (
        [BlogId] int NOT NULL IDENTITY,
        [Description] nvarchar(max) NULL,
        CONSTRAINT [PK_Blogs] PRIMARY KEY ([BlogId])
    );
END;

GO

IF NOT EXISTS(SELECT * FROM [__EFMigrationsHistory] WHERE [MigrationId] = N'20191023140106_MigrationWithMultilineSeedingMssql')
BEGIN
    IF EXISTS (SELECT * FROM [sys].[identity_columns] WHERE [name] IN (N'BlogId', N'Description') AND [object_id] = OBJECT_ID(N'[Blogs]'))
        SET IDENTITY_INSERT [Blogs] ON;
    INSERT INTO [Blogs] ([BlogId], [Description])
    VALUES (1, N'This text
    uses
    3 lines!');
    IF EXISTS (SELECT * FROM [sys].[identity_columns] WHERE [name] IN (N'BlogId', N'Description') AND [object_id] = OBJECT_ID(N'[Blogs]'))
        SET IDENTITY_INSERT [Blogs] OFF;
END;

GO

IF NOT EXISTS(SELECT * FROM [__EFMigrationsHistory] WHERE [MigrationId] = N'20191023140106_MigrationWithMultilineSeedingMssql')
BEGIN
    INSERT INTO [__EFMigrationsHistory] ([MigrationId], [ProductVersion])
    VALUES (N'20191023140106_MigrationWithMultilineSeedingMssql', N'3.0.0');
END;

GO
```

To fix this issue, EF Core could implement one of the following fixes:
- Do not indent the script.
- Re-escape string values before appending them to the script.

For the time being, we might be able to implement a dirty hack, that removes the indentation.

// cc @AndriySvyryd

**Comment by AndriySvyryd** [comment]
This is tracked by https://github.com/aspnet/EntityFrameworkCore/issues/15256

**Comment by tarastriska** [comment]
Thank you very much for your investigation

**Comment by lauxjpn** [comment]
We implemented a workaround with #893 that fixes this issue.
